Implements more advanced ILLink/ILCompiler analysis for Type.GetMember

hamarb123 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/94745

Allows narrowing members down by name & by member type.

hamarb123 committed:
Implement masking based on MemberTypes

- Implements masking the required member types based on the MemberTypes parameter

hamarb123 committed:
Update HandleCallAction.cs

hamarb123 committed:
Implement matching by name

- Matching by name is implemented here, including logic to handle the prefix name lookup scheme of GetMember & special handling for constructors

hamarb123 commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -1216,6 +1278,23 @@ private IEnumerable<MultiValue> ProcessGetMethodByName (TypeProxy type, string m
+			// This also prevents warnings in such case, since if we don't set the return value it will be
+			// "unknown" and consumers may warn.
+			if (!foundAny)
+				yield return NullValue.Instance;
```
Can a reviewer confirm if I want this section here, or not, or if I want it somewhere else? I don't know what it does (it's copied from the non WithPrefix version).

hamarb123 commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -1454,5 +1592,15 @@ internal static bool BindingFlagsAreUnsupported (BindingFlags? bindingFlags)
+
+		private partial IEnumerable<SystemReflectionMethodBaseValue> GetMethodsOnTypeHierarchyWithPrefix (TypeProxy type, string name, bool nameIsPrefix, BindingFlags? bindingFlags);
+
+		private partial IEnumerable<SystemTypeValue> GetNestedTypesOnTypeWithPrefix (TypeProxy type, string name, bool nameIsPrefix, BindingFlags? bindingFlags);
```
Should I move all these `WithPrefix` versions next to the non `WithPrefix` versions in both files & make it so the non-`WithPrefix` version just calls the `WithPrefix` version?

hamarb123 commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -1421,6 +1524,41 @@ internal static bool BindingFlagsAreUnsupported (BindingFlags? bindingFlags)
+
+		internal static DynamicallyAccessedMemberTypes GetDynamicallyAccessedMemberTypesFromMemberTypesForNestedTypes (MemberTypes? memberTypes) =>
+			(HasMemberType (memberTypes, MemberTypes.TypeInfo) ? DynamicallyAccessedMemberTypes.PublicNestedTypes | DynamicallyAccessedMemberTypes.NonPublicNestedTypes : DynamicallyAccessedMemberTypes.None) |
+			(HasMemberType (memberTypes, MemberTypes.NestedType) ? DynamicallyAccessedMemberTypes.PublicNestedTypes | DynamicallyAccessedMemberTypes.NonPublicNestedTypes : DynamicallyAccessedMemberTypes.None) |
```
I'm slightly confused with the distinction between `TypeInfo` and `NestedType` - I think they both refer to nested types?
I assume `TypeInfo` would just be for when there's a nested type without `nested` attribute in IL? Otherwise, I have no idea what it is.
Can someone confirm what it means?

hamarb123 commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -1381,6 +1461,29 @@ internal static bool BindingFlagsAreUnsupported (BindingFlags? bindingFlags)
+
+			// Member types that don't affect binding outside InvokeMember (that we don't analyze).
+			const MemberTypes IgnorableMemberTypes =
+				MemberTypes.Custom;
```
I don't think we need to explicitly preserve attributes currently, do we?

hamarb123 commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -394,13 +400,69 @@ GenericParameterValue genericParam
+						requiredMemberTypesMaskFromMemberTypes = GetDynamicallyAccessedMemberTypesFromMemberTypesForMembers (memberTypes);
+						requiredMemberTypes &= requiredMemberTypesMaskFromMemberTypes;
+					}
 
```
Should we exit early here if there's no flags in the `requiredMemberTypes`? This could happen, for example, with `MemberTypes.Custom`.

hamarb123 commented:
Comment so this doesn't get closed - can someone please review this? Thanks!

hamarb123 committed:
Merge branch 'main' into main7

hamarb123 committed:
Implement methods in other files

hamarb123 committed:
Merge branch 'main7' of https://github.com/hamarb123/runtime into main7

hamarb123 committed:
Fix compile errors

hamarb123 committed:
Fix some compile errors and spacing

- We use EndsWith(string) because some platforms don't have the char overload

hamarb123 committed:
Update HandleCallAction.cs

hamarb123 committed:
Update HandleCallAction.cs

hamarb123 committed:
Add some unit tests & fix some code

- Update the `[Kept]` attributes on the unit tests to reflect what we should now expect to be kept
- Add some more unit tests
- Fix missing `| BindingFlags.Static` for the default binding flags

hamarb123 committed:
Fix typo & move ctors to top

hamarb123 committed:
Update MemberUsedViaReflection.cs

hamarb123 committed:
Update MemberUsedViaReflection.cs

hamarb123 committed:
Update some `[Kept]` attributes

- And add another .IsEmpty check like similar APIs have

sbomer reviewed (commented):
Thanks for the contribution!

hamarb123 commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -367,18 +367,24 @@ GenericParameterValue genericParam
+						// Assume a default value for MemberTypes for methods that don't use MemberTypes as a parameter
+						memberTypes = MemberTypes.All;
+						// Assume a default value for BindingFlags for methods that don't use BindingFlags as a parameter (Type.DefaultLookup)
+						bindingFlags = BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance;
```
I will take a look through this next week some time (away currently).

vitek-karas commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -1216,6 +1278,23 @@ private IEnumerable<MultiValue> ProcessGetMethodByName (TypeProxy type, string m
 				yield return NullValue.Instance;
 		}
 
+		private IEnumerable<MultiValue> ProcessGetMethodByNameWithPrefix (TypeProxy type, string methodName, bool nameIsPrefix, BindingFlags? bindingFlags)
```
Do we think we need to recognize the prefix lookups intrinsically? It seems to bring complexity, but the benefit is questionable (How common is it, to use the prefix lookup functionality - I've never heard about it before this PR honestly). I think it would make sense to treat that case as "unsupported" and revert to the "keep all behavior" in that case.

There will probably always be cases which the trimming doesn't fully implement (it just doesn't make sense to invest resources into emulating every little detail of the runtime behavior, even if possible). As such it makes sense to consciously only implement the things which bring enough value. There's cost associated even with good code, we have to maintain it going forward, future us (and others) will re-read this code to understand and so on.

Same logic applies to the various `MemberType` values.

To put it simply - I don't believe there's value in implementing all of the functionality the runtime has. This is not about correctness, it's just about size. Intentionally leaving certain parts as "unrecognized" will not break applications, just make them larger.

hamarb123 replied:
@vitek-karas I use the `MemberType` values, and it was why I made the issue in the first place...

I can remove the prefix lookup code if we believe it overcomplicates things, as I don't think it provides a lot of value.

vitek-karas replied:
@hamarb123 makes sense. I wasn't commenting on the whole change - specifically I was unsure about the prefixes. Removing that would make sense to me - and it should simplify the change (especially the shred code changes).

hamarb123 committed:
Apply some feedback (incomplete)

- Remove prefix versions of APIs
- Remove ignoring of .Custom
- Add missing space

hamarb123 committed:
Update HandleCallAction.cs

- Remove remaining prefix code
- Don't call AddReturnValue

hamarb123 committed:
Merge branch 'main' into main7

vitek-karas commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -394,13 +400,72 @@ GenericParameterValue genericParam
+									AddReturnValue (MultiValueLattice.Top); // Initialize return value (so that it's not autofilled if there are no matching members)
+
+									// search for all the things we want by name
+									if ((requiredMemberTypes2 & (DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)) != 0) {
```
Why not something like `if (name == ".ctor" && ...)` and remove the check above about constructors - would probably simplify this a little bit.

hamarb123 committed:
Update HandleCallAction.cs

- Implement feedback for .ctor
- Remove calls to `AddReturnValue (MultiValueLattice.Top);`, since this returns an array

hamarb123 commented on src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:
```diff
@@ -1381,6 +1461,29 @@ internal static bool BindingFlagsAreUnsupported (BindingFlags? bindingFlags)
+				MemberTypes.Method |
+				MemberTypes.Property |
+				MemberTypes.TypeInfo |
+				MemberTypes.NestedType;
```
I haven't implemented the changes to the tests yet, will close these comments when I do. I want to get the current tests to not fail first, then I'll change them to be what the new code should expect.

hamarb123 committed:
Update HandleCallAction.cs

hamarb123 commented:
> This pull request has been automatically marked `no-recent-activity` because it has not had any activity for 14 days. It will be closed if no further activity occurs within 14 more days. Any new comment (by anyone, not necessarily the author) will remove `no-recent-activity`.

I'm unable to look at this at the moment, or for another few weeks probably. Will re-open or open a new PR when I'm able to if/when I come back to it.
